contributivity_measures.py

```diff
@@ -168,3 +166,704 @@ def not_twice_characteristic(permut):
+    def approx_increment(subset, k, j):
+        return all_models[j][k].predict(make_coordinate(subset, k))[0]
+
+    # sampling
+    t = 0
+    q = -norm.ppf((1 - alpha) / 2, loc=0, scale=1)
+    v_max = 0
+    all_renorms = []
+    all_models = []
+    Subsets = []  # created like this to avoid pointer issue
+    while (
+        t < 100 or t < 4 * q ** 2 * v_max / (sv_accuracy) ** 2
+    ):  # Check if the length of the confidence interval  is below the value of sv_accuracy*characteristic_all_node
+        if t == 0:
+            contributions = np.array([np.zeros(n)])
+        else:
+            contributions = np.vstack((contributions, np.zeros(n)))
+        subsets = []
+        if t % update == 0:  # renew the importance density g
+            j = t // update
+            make_models()
+            # ## compute the renormalization constant of the new importance density for all datatsets
+            renorms = []
+            for k in range(n):
+                list_k = np.delete(np.arange(n), k)
+                renorm = 0
+                for length_combination in range(len(list_k) + 1):
+                    for subset in combinations(
+                        list_k, length_combination
+                    ):  # could be avoided as   prob(np.array(subset))*np.abs(approx_increment(np.array(subset),j)) is constant in the combination
+                        renorm += prob(np.array(subset)) * np.abs(
+                            approx_increment(np.array(subset), k, j)
+                        )
+                renorms.append(renorm)
+            all_renorms.append(renorms)
+
+        # generate the new increments(subset)
+        for k in range(n):
+            u = np.random.uniform(0, 1, 1)[0]
+            cumSum = 0
+            list_k = np.delete(np.arange(n), k)
+            for length_combination in range(len(list_k) + 1):
+                for subset in combinations(
+                    list_k, length_combination
+                ):  # could be avoided as   prob(np.array(subset))*np.abs(approx_increment(np.array(subset),j)) is constant in the combination
+                    cumSum += prob(np.array(subset)) * np.abs(
+                        approx_increment(np.array(subset), k, j)
+                    )
+                    if cumSum / all_renorms[j][k] > u:
+                        S = np.array(subset)
+                        subsets.append(S)
+                        break
+                if cumSum / all_renorms[j][k] > u:
+                    break
+            SUk = np.append(S, k)
+            increment = not_twice_characteristic(SUk) - not_twice_characteristic(S)
+            contributions[t - 1][k] = (
+                increment * all_renorms[j][k] / np.abs(approx_increment(S, k, j))
+            )
+        Subsets.append(subsets)
+        shap = np.mean(contributions, axis=0)
+        # calcul des variances
+        v_max = np.max(np.var(contributions, axis=0))
+        t += 1
+        print(t)
```

Comment:
The prinds floods the output. We can remove the print or print every 1000th iteration for instance ?